includes/Pages/UserAuth/Login/LoginCredentialPageBase.php

```diff
@@ -0,0 +1,335 @@
+                    $this->redirect('login');
+
+                    return;
+                }
+            }
+            catch (OAuthException $ex) {
+                // Oops. Refreshing ticket failed. Force a re-auth.
+                $authoriseUrl = $oauth->getRequestToken();
+                WebRequest::setOAuthPartialLogin($user);
+                $this->redirectUrl($authoriseUrl);
+
+                return;
+            }
+        }
+
+        if (($this->getSiteConfiguration()->getEnforceOAuth() && !$oauth->isFullyLinked())
+            || $oauth->isPartiallyLinked()
+        ) {
+            $authoriseUrl = $oauth->getRequestToken();
+            WebRequest::setOAuthPartialLogin($user);
+            $this->redirectUrl($authoriseUrl);
+
+            return;
+        }
+
+        WebRequest::setLoggedInUser($user);
+
+        $this->goBackWhenceYouCame($user);
+    }
+
+    protected abstract function getProviderCredentials();
+
+    /**
+     * @param User        $user
+     * @param int         $partialStage
+     * @param PdoDatabase $database
+     *
+     * @throws ApplicationLogicException
+     */
+    private function processJumpNextStage(User $user, $partialStage, PdoDatabase $database)
+    {
+        WebRequest::setAuthPartialLogin($user->getId(), $partialStage + 1);
+
+        $sql = 'SELECT type FROM credential WHERE user = :user AND factor = :stage AND disabled = 0 ORDER BY priority';
+        $statement = $database->prepare($sql);
+        $statement->execute(array(':user' => $user->getId(), ':stage' => $partialStage + 1));
+        $nextStage = $statement->fetchColumn();
+        $statement->closeCursor();
+
+        if (!isset($this->nextPageMap[$nextStage])) {
+            throw new ApplicationLogicException('Unknown page handler for next authentication stage.');
+        }
+
+        $this->redirect("login/" . $this->nextPageMap[$nextStage]);
+    }
+
+    private function setupAlternates(User $user, $partialStage, PdoDatabase $database)
+    {
+        // get the providers available
+        $sql = 'SELECT type FROM credential WHERE user = :user AND factor = :stage AND disabled = 0';
+        $statement = $database->prepare($sql);
+        $statement->execute(array(':user' => $user->getId(), ':stage' => $partialStage));
+        $alternates = $statement->fetchAll(PDO::FETCH_COLUMN);
+
+        $types = array();
+        foreach ($alternates as $item) {
+            $type = $this->nextPageMap[$item];
+            if (!isset($types[$type])) {
+                $types[$type] = array();
+            }
+
+            $types[$type][] = $item;
+        }
+
+        $userOptions = array();
+        if (get_called_class() === PageOtpLogin::class) {
+            $userOptions = $this->setupUserOptionsForType($types, 'u2f', $userOptions);
+        }
+
+        if (get_called_class() === PageU2FLogin::class) {
+            $userOptions = $this->setupUserOptionsForType($types, 'otp', $userOptions);
+        }
```

Comment:
Can you reassure me that this will never have both of them be false, and that the other providers don't have classes? (I couldn't find the file)

Reply:
There are instances where this can be false.

LCPB::setupAlternates() is called by setupPartial() to give the user the option of switching between auth providers at a certain stage. This is user-interface only, and currently this only matters if this authentication stage has at least one credential in the set (yubikey, totp, scratch) **and** also a U2F credential. `PageOtpLogin` handles all the former factors, because they're all a simple text box. U2F is a bit more complex and requires JS.

This has the effect of adding a link on the credential request page which says something like "use u2f instead" or "use a token instead".

If neither of these if blocks matches, there's no link to ask the user for an alternate credential.

If the user somehow gets to an auth page for a credential type they don't have enabled, then when they submit the form the auth request will fail because there's nothing to compare against.